store/tikv/2pc_exec.go

```diff
@@ -0,0 +1,535 @@
+
+	var skipPrimary, skipSecondary bool
+	failpoint.Inject("skipKeyReturnOK", func(val failpoint.Value) {
+		valStr, ok := val.(string)
+		if ok && c.committer.connID > 0 {
+			if c.containsPrimaryKey() && actionIsPessimisticLock {
+				logutil.Logger(c.bo.ctx).Warn("pessimisticLock failpoint", zap.String("valStr", valStr))
+				switch valStr {
+				case "pessimisticLockSkipPrimary":
+					skipPrimary = true
+				case "pessimisticLockSkipSecondary":
+					skipSecondary = true
+				}
+			}
+		}
+	})
+
+	var cancel context.CancelFunc
+	c.bo, cancel = bo.Fork()
+	if actionIsPrewrite {
+		c.cancel = cancel
+	}
+
+	if c.containsPrimaryKey() && (actionIsCommit || actionIsCleanup || actionIsPessimisticLock) && !skipPrimary {
+		keys, err := c.handlePrimaryMutation(actionIsCommit)
+		if err != nil || keys == c.mutations.Len() {
+			return errors.Trace(err)
+		}
+	}
+
+	if skipSecondary {
+		return nil
+	}
+
+	if actionIsCommit {
+		c.bo, cancel = NewBackofferWithVars(context.Background(), CommitMaxBackoff, c.committer.txn.vars).Fork()
+		go func() {
+			err := c.handleMutations()
+			cancel()
+			if err != nil {
+				logutil.BgLogger().Debug("2PC async doActionOnBatches",
+					zap.Uint64("conn", c.committer.connID),
+					zap.Stringer("action type", c.action),
+					zap.Error(err))
+				tikvSecondaryLockCleanupFailureCounterCommit.Inc()
+			}
+		}()
+		return nil
+	}
+
+	err := c.handleMutations()
+	cancel()
+	return errors.Trace(err)
+}
+
+func (c *execController) handlePrimaryMutation(actionIsCommit bool) (int, error) {
+	var (
+		batch *batchMutations
+		err   error
+	)
+
+	for {
+		collector := c.getCollector(c.committer.primaryKey, nil)
+		batch, err = collector.Collect()
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		retry, err := c.action.handleSingleBatch(c.committer, c.bo, batch)
+		if err == nil && !retry {
+			break
+		}
+		if !retry {
+			return 0, errors.Trace(err)
+		}
+		if err != nil {
+			err = c.bo.Backoff(BoRegionMiss, err)
+			if err != nil {
+				return 0, errors.Trace(err)
+			}
+		}
+	}
+
+	knobs := &c.committer.testingKnobs
+	if actionIsCommit && knobs.bkAfterCommitPrimary != nil && knobs.acAfterCommitPrimary != nil {
+		knobs.acAfterCommitPrimary <- struct{}{}
+		<-knobs.bkAfterCommitPrimary
+	}
+
+	start, end := batch.mutations.keys[0], batch.mutations.keys[batch.mutations.len()-1]
+	c.filter = func(key []byte) bool {
+		return bytes.Compare(key, start) < 0 || bytes.Compare(key, end) > 0
+	}
+	return batch.mutations.len(), nil
+}
+
+func (c *execController) handleMutations() (err error) {
+	collector := c.getCollector(nil, nil)
+	for {
+		var batch *batchMutations
+		batch, err = collector.Collect()
+		if err != nil {
+			c.handleError(err)
+			break
+		}
+		if batch == nil {
+			break
+		}
+
+		if collector.Finished() {
+			bo, cancel := c.bo.Fork()
+			job := execJob{batch, bo, cancel}
+			c.handleSingleBatch(&job)
+			break
+		}
+
+		if err = c.dispatchBatch(batch); err != nil {
+			break
+		}
+	}
+	c.Wait()
```

Comment:
Will here wait forever?
If dispatch some of the tasks fail, the worker loop run this:

```
	for job := range c.jobCh {
		if !c.terminated() {
			c.handleSingleBatch(&job)
		}
		c.Done()
	}
```

`jobCh` is not closed, and `terminated()` is false?
If there is no job sending to `jobCh`, there is no change to check `terminated`.

Reply:
The job status is tracked by `execController.WaitGroup`, it will be increased before each job dispatched, and will be decreased after finished.

The `execController.jobCh` is closed after `execController.WaitGroup` decreased to 0.